DataObject support for Scout JS

mvilliger opened the pull request:
- Introduce a DataObjectTransformer which adds annotations during TypeScript compile time to all attributes of DataObjects. These annotations contain the data type of the attribute to be retrieved at runtime.
- Add a base class for DataObjects: BaseDoEntity
- Add ID classes in TS (number, string uuid, composite)
- Add a DoRegistry to retrieve all DataObjects known in TS
- Migrate existing DOs to the new base class including decorators.
- Add convenience methods to ajax utility to post and get DataObjects
- Add a DataObject (De)Serializer for TS. It uses the data type annotations added by the compile time transformer and supports:
  - Arrays
  - Maps
  - Sets
  - Records
  - Dates
  - Ids
- Enable experimental decorators for Scout JS
- Allow the unqualified ID deserializer in Java to accept over-qualified instances (a typeName is available event not necessary because the ID is concrete).
- Migrate from deprecated 'type' jQuery ajax property to the new 'method' property.

389583

fschinkel commented on eclipse-scout-core/src/id/CompositeId.ts:
```diff
@@ -0,0 +1,56 @@
+ */
+import {Id, IdParser, ObjectType, scout} from '../index';
+
+export abstract class CompositeId extends Id<Id<any>[], 'scout.CompositeId'> {
```
Is this class intended to be the pendant of `AbstractCompositeId`? If it is wee need to pass the typeName to the implementing class.
Did you consider introducing generics for the value type? The implementing class can then use tupel types which will increase comfort for the developers using these composite ids.

mvilliger replied:
Good point is fixed.

fschinkel commented on eclipse-scout-core/src/id/IdParser.ts:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  protected _splitSignature(value: string): string[] {
+    let splitPos = value.lastIndexOf(IdParser.SIGNATURE_DELIMITER);
```
is this split-logic really necessary? splitting the string will result in an array with length 1 or 2 and there is no such thing as IndexOutOfBounds in JS

mvilliger replied:
I prefer lastIndexOf because it searches from the end while split would search from the start.

fschinkel commented on eclipse-scout-core/test/uinotification/uiNotificationsSpec.ts:
```diff
@@ -92,9 +92,10 @@ describe('uiNotifications', () => {
+      let response = scout.create(UiNotificationResponse, {
         notifications: [{
           id: '1',
+          objectType: UiNotificationDo,
```
Can we eliminate this?

mvilliger replied:
We could yes, but for this fixture it is required to have the same structure as it would come from the deserializer

fschinkel commented on eclipse-scout-core/src/ajax/ajax.ts:
```diff
@@ -192,13 +192,40 @@ export const ajax = {
     return ajax.createCall(opts, model);
   },
 
+  getDataObject<TDoOut extends BaseDoEntity>(url?: string | JQuery.UrlAjaxSettings, model?: AjaxCallModel): JQuery.Promise<TDoOut, AjaxError> {
```
please add js doc

mvilliger replied:
JS doc is completely missing and will be added later on. Same for documentation

fschinkel commented on eclipse-scout-core/test/dataobject/DoDeserializerSpec.ts:
```diff
@@ -0,0 +1,218 @@
+    expect(resultFromConstructorObjectType.nestedNestedDate).toEqual(dates.parseJsonDate('2024-07-25 08:41:10.708Z'));
+  });
+
+  it('Uses BaseDoEntity if no type information is available', () => {
```
uses

mvilliger replied:
Fixed

fschinkel commented on eclipse-scout-core/src/dataobject/DoRegistry.ts:
```diff
@@ -0,0 +1,94 @@
+      .forEach(doClass => this.add(doClass));
+  }
+
+  add(DoClass: Constructor<BaseDoEntity>, jsonType?: string) {
```
doClass, also see other methods

mvilliger replied:
lowercase is correct here

fschinkel commented on eclipse-scout-core/src/dataobject/serialize/doValueMetaData.ts:
```diff
@@ -0,0 +1,78 @@
+   * Checks if `actualType` is instanceof `declaredType`.
+   */
+  assertTypesCompatible(actualType: Constructor, declaredType: Constructor) {
+    if (actualType && declaredType && actualType !== declaredType && !declaredType.isPrototypeOf(actualType)) {
```
flip if?

mvilliger replied:
Would be possible yes, but makes it not really better: one more line vs one more indent

fschinkel commented on eclipse-scout-core/src/dataobject/serialize/IdDoNodeSerializer.ts:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    // Type name cannot be computed. Check if ID type is fixed on instance (e.g. using the idTypeName annotation or other constructor implementation)
+    const IdType = metaData.type;
```
idType

mvilliger replied:
Capitalized because a new instance is created. necessary to make eslint happy.

fschinkel commented on eclipse-scout-core/src/dataobject/serialize/ArrayDoNodeSerializer.ts:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  static getArrayValueType(metaData: DoValueMetaData): DoValueMetaData {
+    if (metaData?.args?.length) {
```
```suggestion
return metaData?.args?.[0];
```

mvilliger replied:
Is applied

fschinkel commented on eclipse-scout-core/src/dataobject/dataObjects.ts:
```diff
@@ -7,10 +7,54 @@
 export const dataObjects = {
+
+  serializers: [
+    new DateDoNodeSerializer(),
```
how do other layers/projects contribute new serializers? do they simply inject them into this array at the correct position?

mvilliger replied:
Yes

cguglielmo commented on eclipse-scout-core/src/dataobject/serialize/DoDeserializer.ts:
```diff
@@ -0,0 +1,77 @@
+      // use custom deserializer
+      return deserializer.deserialize(value, valueMetaData, this);
+    }
+    if (objects.isNullOrUndefined(value)) {
```
Shouldn't this be the first statement? Or are there serializers that can serialize null or undefined?

mvilliger replied:
No, it is intended like this. If the customizable deserializers come first, it gives them the power to e.g. replace an undefined or null value with e.g. an empty string if an object may require that.
If the null/undefined case would be handled first, such flexibility would get lost.

cguglielmo commented on eclipse-scout-core/src/id/IdParser.ts:
```diff
@@ -0,0 +1,94 @@
+    return this.fromUnqualified(idObjectType, value, typeNameProvider());
+  }
+
+  fromUnqualified<TIdType extends Id<TValueType, TTypeName>, TTypeName extends string, TValueType>(idObjectType: ObjectType<TIdType>, value: string, typeName: string): TIdType {
```
is it necessary to set ensureUniqueId: false ? ObjectFactory excludes Id type

mvilliger replied:
No it's not necessary. Its just another layer of security in case we don't want to keep the change in the ObjectFactory.

cguglielmo commented on eclipse-scout-core/src/id/NumberId.ts:
```diff
@@ -0,0 +1,39 @@
+    if (typeof value === 'number') {
+      this.value = value;
+    } else {
+      const invalidNumberId = () => new Error('Invalid NumberId: ' + arrays.format(value, IdParser.COMPONENT_DELIMITER));
```
maybe extract to method instead of creating function on the fly?

mvilliger replied:
Agree. Fixed.

cguglielmo commented on eclipse-scout-core/src/id/StringId.ts:
```diff
@@ -0,0 +1,36 @@
+    if (typeof value === 'string') {
+      this.value = value;
+    } else {
+      const invalidStringId = () => new Error('Invalid StringId: ' + arrays.format(value, IdParser.COMPONENT_DELIMITER));
```
see NumberId comment

mvilliger replied:
Agree. Fixed.

cguglielmo commented on eclipse-scout-core/src/id/UuId.ts:
```diff
@@ -0,0 +1,36 @@
+    if (typeof value === 'string') {
+      this.value = value;
+    } else {
+      const invalidUuId = () => new Error('Invalid UuId: ' + arrays.format(value, IdParser.COMPONENT_DELIMITER));
```
see NumberId comment, maybe move to Id to reuse it

mvilliger replied:
Agree. Fixed.

cguglielmo commented on eclipse-scout-core/test/dataobject/dataObjectsSpec.ts:
```diff
@@ -23,10 +23,10 @@ describe('dataObjects', () => {
   describe('addContribution', () => {
     it('adds a contribution to the DO', () => {
-      let doEntity: DoEntityWithContributions = {};
+      const doEntity: DoEntityWithContributions = {};
```
we don't have a guideline for let vs const (yet). No need to replace let with const everywhere. I personally prefer let over const

mvilliger replied:
Ok. fine with me. Especially in specs for me const is useful to not by accident overwrite fixture variables from previous expectations.

bschwarzent commented on eclipse-scout-core/src/ObjectFactory.ts:
```diff
@@ -164,10 +164,11 @@ export class ObjectFactory {
     // Create object
-    let scoutObject = this._createObjectByType(objectType, options);
+    const scoutObject = this._createObjectByType(objectType, options);
+    const ensureUniqueId = scout.nvl(options.ensureUniqueId, !(scoutObject instanceof BaseDoEntity || scoutObject instanceof Id) /* don't create unique IDs for DOs and IDs by default */);
```
Maybe extract this logic (or the actual assignment of the `id`) to a method, so it can be customized?

mvilliger replied:
Agree. Fixed

bschwarzent commented on eclipse-scout-core/src/ObjectFactory.ts:
```diff
@@ -242,6 +243,24 @@ export class ObjectFactory {
+   * @returns All classes that have the given class in their super hierarchy. The given baseClass is not part of the result.
+   * More formally: all constructors known to this factory that have the given class in their prototype chain.
+   */
+  getSubClassesOf<T extends object>(baseClass: Constructor<T> | AbstractConstructor<T>): Constructor<T>[] {
```
Is `extends object` necessary?

mvilliger replied:
Basically no as it already must be a constructor. But it is the same as e.g. in the create function and makes it more clear that it makes no sense for primitives.

bschwarzent commented on eclipse-scout-core/src/ObjectFactory.ts:
```diff
@@ -242,6 +243,24 @@ export class ObjectFactory {
+    if (!baseClass) {
+      return result;
+    }
+    for (let obj of this._objectTypeMap.keys()) {
```
Can the type of `_objectTypeMap` be simplified to `Map<Constructor, string>`?

mvilliger replied:
Yes. Fixed.

bschwarzent commented on eclipse-scout-core/src/ajax/ajax.ts:
```diff
@@ -192,13 +192,40 @@ export const ajax = {
     return ajax.createCall(opts, model);
   },
 
+  getDataObject<TDoOut extends BaseDoEntity>(url?: string | JQuery.UrlAjaxSettings, model?: AjaxCallModel): JQuery.Promise<TDoOut, AjaxError> {
```
Why can `url` also be an UrlAjaxSettings object? The other `get*` methods in this utility only use `string`. I would use the same pattern here. If the caller wants to set other options, they should use `callDataObject`.

mvilliger replied:
Agree. Fixed.

bschwarzent commented on eclipse-scout-core/src/ajax/ajax.ts:
```diff
@@ -192,13 +192,40 @@ export const ajax = {
+    return ajax.callDataObject(null, opts, model);
+  },
+
+  postDataObject<TDoIn extends BaseDoEntity | void, TDoOut extends BaseDoEntity | void>(dataObject: TDoIn, url?: string | JQuery.UrlAjaxSettings, model?: AjaxCallModel): JQuery.Promise<TDoOut, AjaxError> {
```
We should probably offer the same convenice functions for "DataObject" as for "Json" -> add `putDataObject` and `removeDataObject`.

mvilliger replied:
Agree. Fixed.

bschwarzent commented on eclipse-scout-core/src/ajax/ajax.ts:
```diff
@@ -192,13 +192,40 @@ export const ajax = {
+  },
+
+  createCallDataObject(dataObject: DoEntity | void, urlAjaxSettings?: JQuery.UrlAjaxSettings, model?: AjaxCallModel): AjaxCall {
+    const json = dataObject ? dataObjects.stringify(dataObject) : undefined;
```
"undefined" is only used to omit the property in the result of the extend() call. I would simplify this here (`json = dataObjects.stringify()`) and use `data: json || undefined` below instead.

mvilliger replied:
Ok. Changed.

bschwarzent commented on eclipse-scout-core/src/code/CodeTypeCache.ts:
```diff
@@ -189,14 +190,52 @@ export class CodeTypeCache extends EventEmitter implements ObjectModel<CodeTypeC
 }
 
+@typeName('scout.Code')
+export class CodeDo extends BaseDoEntity implements CodeModel<any> {
```
Is this defined in the correct file? Why not in Code.ts? Or even in a dedicated file like we do for widget models.

mvilliger replied:
Well, the class is actually only used in the cache. That's why it is declared there. But I extracted it into own files. Then the cache has less noise.

bschwarzent commented on eclipse-scout-core/src/dataobject/DoEntity.ts:
```diff
@@ -8,12 +8,51 @@
+
 /**
- * Base interface for all data object entities.
+ * Base interface for all data objects.
```
Encourage readers to extend the BaseDoEntity class instead of implementing this interface. This mainly exists for legacy reasons, right?

mvilliger replied:
Yes, JsDoc is completely missing and needs to be added/updated anyway.

bschwarzent commented on eclipse-scout-core/src/dataobject/DoEntity.ts:
```diff
@@ -8,12 +8,51 @@
 }
+
+export class BaseDoEntity implements DoEntity, BaseDoEntityModel {
+  declare model: Partial<this> | BaseDoEntityModel;
```
- What is the intention of "Partial<this>"?
- Why is the model argument in the init() method typed as "any" instead of InitModelOf<this>? What other options are there? init() does nothing if the argument is not a POJO. Assertion?

mvilliger replied:
- Partial ensures all properties are optional in the model. In case a property in the DO is not.
- Argument changed to InitModelOf<this>. I would not add an assertion because if a subclass would like to change that behavior, it cannot make a super call anymore. Futhermore the model may also be null which is correct.

bschwarzent commented on eclipse-scout-core/src/dataobject/DoRegistry.ts:
```diff
@@ -0,0 +1,94 @@
+      .forEach(doClass => this.add(doClass));
+  }
+
+  add(DoClass: Constructor<BaseDoEntity>, jsonType?: string) {
```
Is it really that useful to use uppercase for the `DoClass` argument, just because it happens to be of type "Constructor"? I would reserve uppercase names for things that are actually classes.

mvilliger replied:
Well, it is used in new (e.g. in _readJsonType). And ESLint compains if lowerCase variables are used with new. But I limited the uppercase variables to the minimum now.

bschwarzent commented on eclipse-scout-core/src/dataobject/DoRegistry.ts:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    this._constructorByJsonType.set(jsonType, DoClass);
+    const objectType = ObjectFactory.get().getObjectType(DoClass);
```
General remark: should we really default to "const" instead of "let" for local variables? Or is that just the default inserted by the IDE? I fail to see the benefit of defining everything as const. It is similar to the "final" Keyword in Java which we used excessively at some point in time but have since stopped using.

mvilliger replied:
I prefer const over let to ensure I don't override the variable by accident. But it is longer. I think in the end it doesn't matter. As a reader of the code const helps me to know, that this variable will not be changed anywhere below which helps to follow the code flow in my opinion. The difference to Java is that a variable can be declared without keyword while in JS one keyword is required anyway. So I prefer the one which explains the variable better. But it's no strong opinion towards const.

bschwarzent commented on eclipse-scout-core/src/dataobject/DoRegistry.ts:
```diff
@@ -0,0 +1,94 @@
+    return new DoClass()._type;
+  }
+
+  allDataObjects(): IterableIterator<Constructor<BaseDoEntity>> {
```
Check if there is a better name. The method does not really return data objects (only data object _classes_).

mvilliger replied:
Agree. Fixed.

bschwarzent commented on eclipse-scout-core/src/dataobject/serialize/ArrayDoNodeSerializer.ts:
```diff
@@ -0,0 +1,35 @@
+    return array.map(e => deserializer.deserialize(e, arrayValueType));
+  }
+
+  static getArrayValueType(metaData: DoValueMetaData): DoValueMetaData {
```
More suitable in doValueMetaData.ts?

mvilliger replied:
Doesn't really matter. I changed it to the utility.

bschwarzent commented on eclipse-scout-core/src/dataobject/serialize/DoDeserializer.ts:
```diff
@@ -0,0 +1,74 @@
+
+    let constructor = metaData?.type as Constructor<T>;
+    if (constructor && detectedClass) {
+      // if both is present: check compatibility and use the one from _type attribute (might be a subclass of the one in the source code)
```
"both is present" -> "both are present"

mvilliger replied:
Correct. Fixed.

bschwarzent commented on eclipse-scout-core/src/dataobject/serialize/DoDeserializer.ts:
```diff
@@ -0,0 +1,74 @@
+      constructor = BaseDoEntity as Constructor<T>;
+    }
+
+    const resultObj = scout.create(constructor, null /* must always be possible to create a DO without model */, {ensureUniqueId: false});
```
Is the "ensureUniqueId" argument still needed? ObjectFactory will automatically skip id assignment if object is a BaseDoEntity.

mvilliger replied:
No, it's not needed. It was only there in case we don't want to keep the new rule in the ObjectFactory. But as this rule seems to be fine, i removed the option here already.

bschwarzent commented on eclipse-scout-core/src/dataobject/serialize/DoSerializer.ts:
```diff
@@ -0,0 +1,139 @@
+      return;
+    }
+    if (this.contains(value)) {
+      const path = '[' + strings.join(',', ...this._attributeNames) + ']';
```
`this._attributeNames.join(',')` ?

mvilliger replied:
Yes, fine with me. But is the behavior of the native join function really the same as the one from Scout?
Especially because the scout version uses strings.asString which has some extra logic?

bschwarzent commented on eclipse-scout-core/src/dataobject/serialize/DoSerializer.ts:
```diff
@@ -0,0 +1,139 @@
+    return result;
+  }
+
+  protected _setJsonTypeTo(obj: any, constructor: Constructor) {
```
`_ensureJsonType` ?

mvilliger replied:
Agree. Fixed.

bschwarzent commented on eclipse-scout-core/src/dataobject/serialize/IdDoNodeSerializer.ts:
```diff
@@ -0,0 +1,78 @@
+    return IdDoNodeSerializer._getIdParser().parse(metaData.type, rawId, () => this.detectConcreteTypeName(metaData));
+  }
+
+  isIdClass(candidate: Constructor) {
```
Missing return type.

Also: what is the expected ts-error here?

mvilliger replied:
Correct. Fixed.

The ts error is TS2367: unintentional comparison.

bschwarzent commented on eclipse-scout-core/src/dataobject/serialize/IdDoNodeSerializer.ts:
```diff
@@ -0,0 +1,78 @@
+  }
+
+  parseTypeName(metaData: DoValueMetaData): string {
+    return metaData.type['parseTypeName'](metaData);
```
Why is bracket notation necessary here? Is this not part of the official API?

mvilliger replied:
It is part of the API. But the call is by reflection here. Is there a better way to write this? As this method could be overwritten.

bschwarzent commented on eclipse-scout-core/src/dataobject/serialize/IdDoNodeSerializer.ts:
```diff
@@ -0,0 +1,78 @@
+
+export class IdDoNodeSerializer implements DoNodeSerializer<Id<any>> {
+
+  protected static _ID_PARSER: IdParser; // created on first use. Access using IdDoNodeSerializer._getIdParser()
```
Alternatively, we could provide a getSingleton() method on the IdParser class itself.

mvilliger replied:
Agree. Fixed.

bschwarzent commented on eclipse-scout-core/src/ajax/ajax.ts:
```diff
@@ -26,7 +26,7 @@ export const ajax = {
     let opts = $.extend({}, {
       url: url,
-      type: 'GET'
+      method: 'GET'
```
Consider a separate commit for this change, it has nothing to do with data objects.

mvilliger replied:
That's true. But the extra work to separate this now does not pay out here I think.

bschwarzent commented on eclipse-scout-core/src/id/Id.ts:
```diff
@@ -0,0 +1,111 @@
+    return this.toUnqualified();
+  }
+
+  static parseTypeName(idMetaData: DoValueMetaData): string {
```
No parsing takes place in this function. Is it even called somewhere?

mvilliger replied:
Right. I changed the name. Yes, it is used by reflection in IdDoNodeSerializer

bschwarzent commented on eclipse-scout-core/src/id/Id.ts:
```diff
@@ -0,0 +1,111 @@
+      return true;
+    }
+
+    if (a instanceof Date && b instanceof Date) {
```
This seems odd and should at least be explained. Do we really use Ids that contain a Date? Shouldn't objects.equals() already handle this?

mvilliger replied:
Yes, we really have IDs with dates. This is officially supported and used.

Yes objects.equals SHOULD handle this, but it does not. And we don't think that this is a good moment to change that. After discussions with cgu about the same topic, we decided to introduce an equals2 and equalsRecursively2 which handles these cases better. In the next release it is the plan to switch to the new implementation so that we have time to stabilize again.

bschwarzent commented on eclipse-scout-core/src/id/IdParser.ts:
```diff
@@ -0,0 +1,94 @@
+
+export class IdParser implements ObjectWithType {
+
+  static ID_TYPENAME_DELIMITER = ':';
```
What does this constant have the ID_ prefix, but the other two don't? Maybe this should be called TYPE_NAME_DELIMITER?

mvilliger replied:
Agree. Fixed.

bschwarzent commented on eclipse-scout-core/src/id/Id.ts:
```diff
@@ -0,0 +1,111 @@
+    this._initIdValue(value);
+  }
+
+  protected abstract _initIdValue(value: TIdValueType | string[]);
```
When is the argument a string array? Is this only needed for the CompositeId? But is the value of the composite id not already an array?

mvilliger replied:
See e.g. DomainKey.ts
The value type of the ID (type arg) can be different than the serialized model.

mvilliger committed:
DataObject support for Scout JS

- Introduce a DataObjectTransformer which adds annotations during
  TypeScript compile time to all attributes of DataObjects. These
  annotations contain the data type of the attribute to be retrieved at
  runtime.
- Add a base class for DataObjects: BaseDoEntity
- Add ID classes in TS (number, string uuid, composite)
- Add a DoRegistry to retrieve all DataObjects known in TS
- Migrate existing DOs to the new base class including decorators.
- Add convenience methods to ajax utility to post and get DataObjects
- Add a DataObject (De)Serializer for TS. It uses the data type
  annotations added by the compile time transformer and supports:
  - Arrays
  - Maps
  - Sets
  - Records
  - Dates
  - Ids
- Enable experimental decorators for Scout JS
- Allow the unqualified ID deserializer in Java to accept over-qualified
  instances (a typeName is available event not necessary because the ID
  is concrete).
- Migrate from deprecated 'type' jQuery ajax property to the new
  'method' property.
- Add objects.equals2 and equalsRecursive2. Differences to original:
  - Dates are supported (compared by value)
  - Classes must be same for two objects to be equal
  - empty Sets/Maps are considered equal
- Add objects.equalsSet & objects.equalsMap
  In a future release these new functions will replace the existing
  ones.

389583

cguglielmo commented:
Merged with #1413
